### DISCUSSION TITLE
List of Continuous, Non-Contractual `clv` Models to Consider Adding

### DISCUSSION THREAD

**Discussion by ricardoV94** [pymc-labs/pymc-marketing#1230] [Announcements]
https://github.com/mplatzer/BTYDplus

**Comment by ColtAllen** [comment]
### Model List

1.  **Modified BG/NBD**: https://github.com/pymc-labs/pymc-marketing/issues/94
2. **BG/CNBD-k**: http://www.reutterer.com/papers/reutterer&platzer&schroeder_2021.pdf
3. **Modified )BG/CNBD-k**: http://www.reutterer.com/papers/reutterer&platzer&schroeder_2021.pdf
4. **Abe variant of Pareto/NBD**: [[1]](https://core.ac.uk/download/pdf/18510752.pdf),[[2]](https://link.springer.com/article/10.1007/s11573-021-01057-6),[[3]](http://www.cirje.e.u-tokyo.ac.jp/research/dp/2008/2008cf537.pdf),[[4]](http://www.cirje.e.u-tokyo.ac.jp/research/dp/2009/2009cf616.pdf)
5. **Pareto/GGG**: http://www.reutterer.com/papers/platzer&reutterer_pareto-ggg_2016.pdf
6. **Gamma/Gompertz/NBD**: free PDF available
7. **Gamma/Gompertz/NBD w/ covariates**: https://pubsonline.informs.org/doi/epdf/10.1287/mksc.2020.1254
8.  **Pareto/NBD w/ time-varying covariates**: https://www.zora.uzh.ch/id/eprint/197236/1/mksc.2020.1254.pdf
9. **BG/NBD w/time-varying covariates**: https://www.brucehardie.com/notes/040/time-varying_covariates_in_BGNBD.pdf
10. **Periodic Death Opportunity**: [[1]](https://repository.upenn.edu/cgi/viewcontent.cgi?article=1343&context=fnce_papers) & [[2]](https://core.ac.uk/download/pdf/18510752.pdf)
11. **Gaussian Processes**: https://www0.gsb.columbia.edu/mygsb/faculty/research/pubfiles/26107/ansari_Bayesian_Nonparametric.pdf

### Notes on Implementation

- The papers for Models 4 & 5 provide MCMC sampling strategies.
- Models 5 & 9 require raw transactions as inputs. All other models follow the frequency/recency/T format.
- Should cite the R libraries where appropriate: https://github.com/mplatzer/BTYDplus & https://github.com/bachmannpatrick/CLVTools
- On that note, [CLVTools contains C++ code for model 8](https://github.com/bachmannpatrick/CLVTools/blob/master/src/pnbd_dyncov_LL.cpp). Consider adapting those scripts with a python binding, because this model is by far the most complex on this list. The likelihood function alone sprawls across 2 pages in the research paper! 
- Models 6 & 7 utilize a Gompertz distribution, which is not currently provided in `pymc`. The likelihood function also contains an infinite integral, which I'm not sure how to code out in `pytensor`.
- The papers for models 5 & 7 cite even more options we could add to this list.
- The paper for model 9 also includes a section on [incorporating seasonality which could implemented in any model](https://github.com/pymc-labs/pymc-marketing/issues/219), which I would prefer to do in lieu of model 8.

**Comment by zwelitunyiswa** [comment]
@ColtAllen The Gamma/Gompertz/NBD model with time-invariant contextual factors (Näf, Bachmann & Meierer 2020) paper is here: https://pubsonline.informs.org/doi/epdf/10.1287/mksc.2020.1254

I also have the Standard Gamma/Gompertz/NBD (Bemmaor & Glady 2012) paper. I can give it to you via email or DM.

**Comment by zwelitunyiswa** [comment]
@ColtAllen Check your LinkedIn inbox.

**Comment by ColtAllen** [comment]
Thanks @zwelitunyiswa!

**Comment by ColtAllen** [comment]
Pinning this because it's a good list of models to add to the `clv` module and research reference. Also added a link to another paper comparing the Pareto/NBD, BetaGeo/NBD, PDO and Bayesian Pareto/NBD Abe models.

**Comment by larryshamalama** [comment]
There's a lot of models that seem available, each with a different paper. Perhaps it would be good to find a way to harmonize all the models in a table or some sort of documentation? On top of other models (discrete, contractual, etc.)

**Comment by ColtAllen** [comment]
> There's a lot of models that seem available, each with a different paper. Perhaps it would be good to find a way to harmonize all the models in a table or some sort of documentation? On top of other models (discrete, contractual, etc.)

That's a good idea. Do you think it's worth adding to the official documentation?

**Comment by larryshamalama** [comment]
> That's a good idea. Do you think it's worth adding to the official documentation?

If I can find a clever and intuitive way of summarizing the models 😅 Something that I will have to think about...

**Comment by ColtAllen** [comment]
> > That's a good idea. Do you think it's worth adding to the official documentation?
> 
> If I can find a clever and intuitive way of summarizing the models 😅 Something that I will have to think about...

I edited my OP so it's more coherent. For the official documentation, use case recommendations would be the best way to summarize these models methinks, preferably after some real-world application.

**Comment by ColtAllen** [comment]
Updated OP to add issue link for generalized seasonality.

**Comment by ColtAllen** [comment]
Converted to discussion and updated issue for [Modified Beta Geo Model](https://github.com/pymc-labs/pymc-marketing/issues/94).

**Comment by PabloRoque** [comment]
Giving [1] a go in #1301
